docs/extensions.rst

```diff
@@ -133,3 +133,38 @@ TLS
 ---
 
 See :doc:`specs/tls`.
+
+HTTP Trailers
+---
+
+The Trailer response header allows the sender to include additional fields at the
+end of chunked messages in order to supply metadata that might be dynamically
+generated while the message body is sent, such as a message integrity check,
+digital signature, or post-processing status.
+
+ASGI servers that implement this extension will provide
+``http.response.trailers`` in the extensions part of the scope::
+
+    "scope": {
+        ...
+        "extensions": {
+            "http.response.trailers": {},
+        },
+    }
+
+An ASGI framework interested in sending trailing headers to the client, must include
+the ``Trailer`` header in the ``http.response.start``. That will allow the ASGI server
```

Comment:
Is the ``trailer`` field in ``http.response.start`` necessary for server implementations？I didn't have that field but nonecorn still works.

Reply:
How do you make sure that the server will not close the transport without informing that the trailing headers will be available? Making the _last_ `http.response.body` to send `more_body=True`? 🤔

Reply:
>  Making the last http.response.body to send more_body=True

Yes, this is exactly what I am doing.

> I didn't add the trailer field in the http.response.start on this spec. I've mentioned that the Trailer header should be included in the headers field on the http.response.start.

I saw it from [uvicorn's pr](https://github.com/encode/uvicorn/pull/1599/files) which seems to have a trailer field.

Reply:
> I saw it from [uvicorn's pr](https://github.com/encode/uvicorn/pull/1599/files) which seems to have a trailer field.

I think you didn't see it, but I mentioned on the description of this PR:

> The uvicorn one introduces a new key on the `http.response.start` called `trailers`, to show the intention from the application to the server that trailing headers will be sent. I think this is not needed. The server can understand the intention if the header `Trailing:` is provided on the `headers` from the `http.response.start`.

Reply:
> Yes, this is exactly what I am doing. And add a filed for that event, like { 'type': 'http.response.trailers', 'trailers': [ [b'x-trailer-test', b'test'], ], "more_body": False }

Got it. But then... The ASGI framework still needs to send the `Trailer` header. So the `more_body` set to `False` on the last `http.response.body` is also not needed, because the server already knows about the `Trailer` header.

Reply:
Actually I didn't look into the headers when I wrote that... So you mean we can improve it by letting server automatically judge from the headers? I just let asgi applications send as they want, no need to add extra headers.

Reply:
Well, according to the RFC, your implementation is more accurate. I will fix that soon. But what will happen if an asgi application wants to send trailers multiple times?

Reply:
According to https://www.rfc-editor.org/rfc/rfc2616.html#section-14.40:

>  An HTTP/1.1 message SHOULD include a Trailer header field in a
   message using chunked transfer-coding with a non-empty trailer. Doing
   so allows the recipient to know which header fields to expect in the
   trailer.

So it should be included... So I guess, we need to make the server know at the beginning of if it's going to send trailing headers.

Reply:
> Well, according to the RFC, your implementation is more accurate. I will fix that soon. But what will happen if an asgi application wants to send trailers multiple times?

By this spec, it's not allowed. 🤔 Should it be? By this spec, I'd expect a single `http.response.trailers` message with all the headers in the `headers` (or we can call it `trailers` if wanted) field.

Reply:
I think more than one ```http.response.trailers``` should be possible for some cases, and it's not very hard for server  implementations.

Reply:
I was writing the changes that I would have to implement if this is the case... So I wrote this:

> If I adjust the spec to allow that, then I'll have to add the following changes in the document:
> - The last `http.response.body` will send `more_body` as `True`.
> - Add `more_body` field to `http.response.trailers`. 
> - The last `http.response.trailers` will send `more_body` as `False`.
> Just to be clear, the ASGI application will still be in charge of sending the `Trailer:` header.

But then I noticed... What if the `Trailer:` header is not sent by the ASGI application? Does the ASGI server waits for all `http.response.trailers`, and does not send those to the client? Does the server closes the transport and then raises an error if `http.response.trailers` are detected without the initial `Trailer:` header?

Reply:
Does the ASGI application need to ensure that the data is sent as required, or does the server have the obligation to check?

Reply:
If the check is left to the ASGI app, the implementation of the server will be much easier, and we can build a prototype server from here. After all, the server is called by the ASGI application. If the ASGI application runs fine, there will be no problem. The server only needs to check the necessary fields for sending data.

Reply:
> The server only needs to check the necessary fields for sending data.

The server SHOULD let the client knows the trailing headers it will be sending, as I've mentioned above:

> An HTTP/1.1 message SHOULD include a Trailer header field in a
message using chunked transfer-coding with a non-empty trailer. Doing
so allows the recipient to know which header fields to expect in the
trailer.

I think you're right, the server should not care if the trailing headers keys were sent in the `http.response.start`. But... The ASGI application SHOULD send them.

About the "multiple `http.response.trailers` messages": If that's allowed, then we need a field to allow the server to understand it has more. I guess `more_body` it's the candidate field here...